### ISSUE TITLE
[Compilation Error] `dprintf` argument mismatch

### ISSUE THREAD

**Issue by incognitte** [axeld/driveencryption#5]
Compiling on [Haiku OS R1/beta3 (Revision hrev55181+62) x86_gcc2](https://www.haiku-os.org/get-haiku/r1beta3/) fails with the following error.

> /boot/home/driveencryption/lib/crypt.cpp:1442: passing `const char *' to argument 1 of `dprintf(int, const char *, ...)' lacks a cast
> /boot/system/develop/headers/posix/stdio.h:174: too few arguments to function `int dprintf(int, const char *, ...)'
> /boot/home/driveencryption/lib/crypt.cpp:1442: at this point in file
> /boot/system/develop/etc/makefile-engine:300: recipe for target 'objects.x86-cc2-release/crypt.o' failed
> make[1]: *** [objects.x86-cc2-release/crypt.o] Error 1
> make[1]: Leaving directory '/boot/home/driveencryption/lib'
> makefile:3: recipe for target 'default' failed
> make: *** [default] Error 2

A closer inspection suggests that `dprintf` in [`posix/stdio.h`](https://raw.githubusercontent.com/haiku/haiku/r1beta3/headers/posix/stdio.h) expects file handle as the [first argument](https://github.com/haiku/haiku/blob/c8394c81112cd959a35fca4beb72e4a8449f7412/headers/posix/stdio.h#L174)
```C++
extern int		dprintf(int fd, char const *format, ...) __PRINTFLIKE(2,3);
```
while `crypt.cpp` calls it without this argument ([an example](https://github.com/axeld/driveencryption/blob/c3ecc37605656e4ee3c95dd9a5fbb0b306f00415/lib/crypt.cpp#L456))
```C++
	dprintf("magic: %.4s\n", (char*)&header.magic);
```

I suspect this error is caused after [this commit](https://github.com/haiku/haiku/commit/db9d2a6f14f98f8a23158ae2085b778fcfe8e19b) to Haiku OS

**Comment by axeld** [comment]
Yes, indeed, that makes sense. Looks like we have to define _KERNEL_MODE for the driver then.

**Comment by incognitte** [comment]
I think `_BOOT_MODE` is needed instead of ~`_KERNEL_MODE`~.

If you modify `crypt.cpp` and these lines to [line 9](https://github.com/axeld/driveencryption/blob/master/lib/crypt.cpp#L9)

``` C++
#ifndef _KERNEL_MODE
#	define _KERNEL_MODE
#endif
```
then compilation fails with this error

> [/boot/system/develop/headers/private/kernel/util/kernel_cpp.h:14](https://github.com/haiku/haiku/blob/r1beta3/headers/private/kernel/util/kernel_cpp.h#L14): parse error

however `_BOOT_MODE` compiles properly.

``` C++
// issue #5  dprintf argument mismatch
#ifndef __BOOT_MODE
#	define __BOOT_MODE
#endif
```

**BUT** once the package is compiled and binaries are installed using generating `hpkg`, Haiku OS won't boot anymore and the boot process hangs. This needs more log collection and root cause analyseson its own bug tracker.

**Comment by incognitte** [comment]
In order to get `make test` compile, `_KERNEL_MODE` needs to be defined `0`

However link fails looking for a library called `kernelland_emu`

```make
/storage/projects/driveencryption> make test
make[1]: Entering directory '/storage/projects/driveencryption/test'
cc -o "objects.x86-cc2-release/crypt_test"      objects.x86-cc2-release/crypt_test.o  -Xlinker -soname=_APP_  -L./  ../lib/objects.x86-cc2-release/libcrypt.a  -lkernelland_emu 
/packages/gcc-2.95.3_2017_07_20-2/.self/develop/tools/i586-pc-haiku/bin/ld: cannot find -lkernelland_emu
collect2: ld returned 1 exit status
/boot/system/develop/etc/makefile-engine:249: recipe for target 'objects.x86-cc2-release/crypt_test' failed
make[1]: *** [objects.x86-cc2-release/crypt_test] Error 1
make[1]: Leaving directory '/storage/projects/driveencryption/test'
makefile:20: recipe for target 'test' failed
make: *** [test] Error 2
```

I am now puzzled because, I can find it in the [Haiku source tree](https://git.haiku-os.org/haiku/tree/src/tests/add-ons/kernel/kernelland_emu), but `pkgman` cannot find any reference to this library
```bash
/storage/projects/driveencryption> pkgman search kernelland_emu 
No matching packages found.
```

BTW, here is the output of `git diff` if you want to try by yourself.

```diff
diff --git a/lib/crypt.cpp b/lib/crypt.cpp
index f478ced..2d1de81 100644
--- a/lib/crypt.cpp
+++ b/lib/crypt.cpp
@@ -6,6 +6,9 @@
  * Truecrypt License.
  */
+// issue #5  dprintf argument mismatch 
+#ifndef _KERNEL_MODE
+#      define _KERNEL_MODE 1
+#endif
 
 #include "crypt.h"
 

diff --git a/test/crypt_test.cpp b/test/crypt_test.cpp
index 8bfd2d3..ce93ba6 100644
--- a/test/crypt_test.cpp
+++ b/test/crypt_test.cpp
@@ -3,7 +3,7 @@
  * Distributed under the terms of the MIT License.
  */
 
-
+// issue #5  dprintf argument mismatch
+#define _KERNEL_MODE 0
 #include "crypt.h"
 
 #include "random.h"
